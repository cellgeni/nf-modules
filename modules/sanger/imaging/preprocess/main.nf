process IMAGING_PREPROCESS {
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "quay.io/cellgeni/clij2:0.28"

    input:
    tuple val(meta), val(out_folder), path(root_folder), val(image_id), val(index)
    path(psf_folder, stageAs: 'psfs')

    output:
    tuple val(meta), path("${out_folder}/${out_img_name}"), emit: fovs
    path "versions.yml"           , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    out_img_name = "${image_id}.tif"
    """
    process.py run \\
        --root_folder ${root_folder} \\
        --index ${index} \\
        --out_img_name ${out_img_name} \\
        --out_folder ${out_folder} \\
        --psf_folder ${psf_folder} \\
        $args

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        imagepreprocess: \$(process.py version)
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    out_img_name = "${image_id}.tif"
    """
    mkdir -p ${out_folder}
    touch ${out_folder}/${out_img_name}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        imagepreprocess: \$(process.py version)
    END_VERSIONS
    """
}
