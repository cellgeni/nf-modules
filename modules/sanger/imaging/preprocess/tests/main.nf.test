// nf-test test main.nf.test
nextflow_process {

    name "Test Process IMAGING_PREPROCESS"
    script "../main.nf"
    process "IMAGING_PREPROCESS"
    config "./nextflow.config"

    tag "modules"
    tag "modules_sanger"
    tag "imaging"
    tag "imaging/preprocess"

    test("preprocess - hcs - decon_10") {

        when {
            params {
                module_args = "--master_file Index.xml -iterations 10"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    10, 
                    file(params.modules_testdata_base_path + '/TL_STP/TL_STP_RVT-HFF-20220623__2022-06-24T13_50_10-Measurement 1/Images', checkIfExists: true),
                    "hcs_test"
                ]
                input[1] = file('/lustre/scratch127/cellgen/cellgeni/projects/imaging_PSFs/', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess - wsi - decon_10") {

        when {
            params {
                module_args = "--master_file Index.idx.xml -iterations 10"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    2,
                    file(params.modules_testdata_base_path + '/QC/beadplatetest__2021-02-17T16_03_59-Measurement 1/Images', checkIfExists: true),
                    "wsi_test"
                ]
                input[1] = file('/lustre/scratch127/cellgen/cellgeni/projects/imaging_PSFs/', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess - hcs - decon_100") {

        when {
            params {
                module_args = "--master_file Index.xml"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    10, 
                    file(params.modules_testdata_base_path + '/TL_STP/TL_STP_RVT-HFF-20220623__2022-06-24T13_50_10-Measurement 1/Images', checkIfExists: true),
                    "hcs_test"
                ]
                input[1] = file('/lustre/scratch127/cellgen/cellgeni/projects/imaging_PSFs/', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess - wsi - decon_100") {

        when {
            params {
                module_args = "--master_file Index.idx.xml"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    2,
                    file(params.modules_testdata_base_path + '/QC/beadplatetest__2021-02-17T16_03_59-Measurement 1/Images', checkIfExists: true),
                    "wsi_test"
                ]
                input[1] = file('/lustre/scratch127/cellgen/cellgeni/projects/imaging_PSFs/', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess - hcs") {

        when {
            params {
                module_args = "--master_file Index.xml"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    10, 
                    file(params.modules_testdata_base_path + '/TL_STP/TL_STP_RVT-HFF-20220623__2022-06-24T13_50_10-Measurement 1/Images', checkIfExists: true),
                    "hcs_test"
                ]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess - wsi") {

        when {
            params {
                module_args = "--master_file Index.idx.xml"
            }
            process {
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    2,
                    file(params.modules_testdata_base_path + '/QC/beadplatetest__2021-02-17T16_03_59-Measurement 1/Images', checkIfExists: true),
                    "wsi_test"
                ]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("preprocess -  stub") {

        options "-stub"

        when {
            params {
                module_args = "--master_file Index.idx.xml"
            }
            process {
                
                """
                input[0] = [
                    [ id:'test' ], // meta map
                    0,
                    file(params.modules_testdata_base_path + '', checkIfExists: true),
                    "stub_test"
                ]
                input[1] = []
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
