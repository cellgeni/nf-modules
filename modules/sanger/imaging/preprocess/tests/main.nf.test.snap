{
    "preprocess -  stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "stub_test.ome.tif:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ],
                "fovs": [
                    [
                        {
                            "id": "test"
                        },
                        "stub_test.ome.tif:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-28T23:14:23.267661797"
    },
    "preprocess - wsi": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "wsi_test.ome.tif:md5,c99aa6e2c0cfbf08e1060475341184ec"
                    ]
                ],
                "1": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ],
                "fovs": [
                    [
                        {
                            "id": "test"
                        },
                        "wsi_test.ome.tif:md5,c99aa6e2c0cfbf08e1060475341184ec"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-28T23:06:11.272919977"
    },
    "preprocess - hcs": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "hcs_test.ome.tif:md5,9e33f3e2ca45d3e516e0b03bc9f9015a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ],
                "fovs": [
                    [
                        {
                            "id": "test"
                        },
                        "hcs_test.ome.tif:md5,9e33f3e2ca45d3e516e0b03bc9f9015a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,7532d207df3aa8c97abafc64c9788d28"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-04-28T23:01:26.437855516"
    }
}